### DISCUSSION TITLE
Prepass Pixel Shader Output ID?

### DISCUSSION THREAD

**Discussion by eballeste** [bnpr/Malt#244] [Q&A] [answered]
i was looking at the procedural mesh sample and noticed that the mesh shader has a prepass pixel shader output id (vec4) that helps alter dramatically the line render, what is this ID and how does it help influence what gets outlined in the line render stage?

**Comment by pragma37** [comment, marked as answer]
The ID is just a unique identifier. By default, the first channel (ID.x) is a unique number for each object, and the other 3 channels are just 0.

The line_width function can detect id boundaries for each channel:
![imagen](https://user-images.githubusercontent.com/36610999/153674834-61b9703a-e979-4acb-b00d-f311c0bfbc66.png)

It's best to avoid changing the value of the first channel, since it's used for detecting self shadows too.

Keep in mind that, since it's a uvec4, it stores values as unsigned integer numbers, so it doesn't take decimals into account.

**Reply by eballeste** [reply]
I think I understand, in your screenshot you are basically remapping the main object's id into the first channel and then passing the id of the texture paint into the second channel which results in the `line width` doing edge detection for both channels at the same time is this correct?

**Reply by eballeste** [reply]
Actually I'm a little confused as to what that first ID channel is doing. For context, besides creating edges procedurally with noise, I am also trying to create outlines for shadows. I have a floor mesh where I basically recreated your setup shown above but for the second ID channel im using a diffuse node connected to a luma node mapped to the second channel. this results in the floor plane receiving an outline on both the shadow edges and the plane edges. I just want the shadows to be outlined so I first tried to assign a value of 0 to the first channel with no effect. Then in the line width function I tried to assign a value of 0 to the first id boundary width channel, again with no results. What would be the best way to exclude the main object's edges from the line renderer? No matter what I do, the main edges are always affected by the line renderer.

![Screenshot 2022-02-12 005851](https://user-images.githubusercontent.com/6924108/153699464-c170b273-af38-4ae0-904f-1182bf5f65b3.png)

**Reply by pragma37** [reply]
You still have depth (*depth width*) and normal (*normal width*) enabled.  You should set those 2 values to 0.

But even then, the background ID is (0,0,0,0). And your plane ID is (x,1,0,0) in lit areas and (x,0,0,0) in shadow areas. So there's still an ID boundary between the lit areas and the background.

You could simply swap the value of the second ID channel:
```glsl
uvec4(ID.x, 1 - round(a.x), 0, 0)
```
It also may be better to use *diffuse shading* instead of *half diffuse shading* for the ID selection, since *half diffuse*  shadow value is 0.5 and that can round to either 0 or 1.

**Reply by eballeste** [reply]
mil gracias ✌️✌️

**Reply by eballeste** [reply]
after your last example i 100% understand how IDs work 🙏🙏🙏

**Comment by eballeste** [comment]
Hi @pragma37! 

So sorry for the rookie questions but I was trying to recreate this same exercise using the new nodes and I thought I knew how ID's and edge detection worked but I seem to be completely lost. I have a very simple scene with a plane, a cube, and a light source for shadows. In your previous explanation you mentioned that the background, by default, has an ID of (0, 0, 0, 0). 

For my plane, I am trying to set the ID to (x, 0, 0, 0) in the lit areas and (x, 1, 0, 0) in the dark areas but for some reason both the outer and the inner edges of the plane have an outline. 

This was my first attempt with no edges in the inner shadow:
![Screenshot 2022-09-02 205403](https://user-images.githubusercontent.com/6924108/188249490-4f366bc3-dcca-4dba-af93-66edaf1a6d28.png)

This was my second attempt where I was able to get outlines on the shadows but failed to remove the outer outline between the background / plane:
![Screenshot 2022-09-02 202916](https://user-images.githubusercontent.com/6924108/188249123-d1c54e83-4e74-4920-9911-3d2d93cefd00.png)

Do ID's work differently now? I noticed in your new example shaders file you are using a `Pack ID` node which consists of 4 different vec4's that get combined into a single uvec4?? How does that work?

I also couldn't find the Luma node so i'm using a convert RGB to HSV to be able to use the value for shadow edge detection, is this the right approach to take or should I be using something else? 

One last question, and apologies if this is common knowledge, but, is there a way to view the values of each object's Ids? I feel like I'm trying to figure this out in the dark.

**Reply by pragma37** [reply]
IDs work the same, but your setup misses a few things:
https://github.com/bnpr/Malt/discussions/244#discussioncomment-2163649

Pack ID is just an optional node that generates an ID channel value from a color.
Luma was replaced by Grayscale, but HSV works fine as well.

**Reply by eballeste** [reply]
gotcha, thanks! gonna chalk it up to my lack of knowledge on the subject but it seemed to have been caused by the `RGB to HSV` node. 

after setting all of the line width options to 0, (why do I have to do this by the way? how does this affect the outlines in my specific scenario?), I noticed that I was still not getting the desired result, but as soon as I swapped it with the `Grayscale` node, it worked! 

I would have thought that the 3rd Value channel of the HSV output, would have done the same trick?

`RGB to HSV` : `uvec4(a.x, 1 - round(b.z), 0, 0)`
![Screenshot 2022-09-03 120748](https://user-images.githubusercontent.com/6924108/188278968-2ea1f867-5130-44e5-b286-88f75fd2ab3c.png)

`Grayscale` :  `uvec4(a.x, 1 - round(b), 0, 0)`
![Screenshot 2022-09-03 120657](https://user-images.githubusercontent.com/6924108/188279148-c7f9fb83-650d-44aa-9a96-d986a2f8e9ff.png)

**Reply by eballeste** [reply]
noticing a difference between the raw output of the grayscale vs rgb to hsv node which might explain the issue?? but not sure, i tried making the shadows darker and even then I saw no border show up around the shadows:

`raw ouput of Grayscale node`
![Screenshot 2022-09-03 123609](https://user-images.githubusercontent.com/6924108/188280157-c571d276-af53-49e2-a9e6-ca4d690cd40c.png)

`raw ouput of 3rd channel (value?) of RGB to HSV node`
![Screenshot 2022-09-03 123500](https://user-images.githubusercontent.com/6924108/188280158-b59c5818-3587-4aa3-8588-33ff4f029418.png)
